### PR TITLE
gh-112509: Fix keys being present in both required_keys and optional_keys in TypedDict

### PR DESCRIPTION
**Pull request by JelleZijlstra**
* Issue: gh-112509

### CODE REVIEW

**Comment by alicederyn** [comment]
I don't think this change is required; see https://github.com/python/cpython/issues/112509#issuecomment-1831423442

File: Lib/typing.py [hunk 1]
```diff
@@ -2884,8 +2884,12 @@ def __new__(cls, name, bases, ns, total=True):
...
         for base in bases:
             annotations.update(base.__dict__.get('__annotations__', {}))
-            required_keys.update(base.__dict__.get('__required_keys__', ()))
-            optional_keys.update(base.__dict__.get('__optional_keys__', ()))
+            base_required = base.__dict__.get('__required_keys__', ())
+            base_optional = base.__dict__.get('__optional_keys__', ())
+            required_keys.update(base_required)
+            required_keys.difference_update(base_optional)
+            optional_keys.update(base_optional)
+            optional_keys.difference_update(base_required)
```
Comment (AlexWaygood) [inline Lib/typing.py]: It took me a minute or two to wrap my head around this. I'd find this a little more readable:

```suggestion

            base_required = base.__dict__.get('__required_keys__', set())
            required_keys |= base_required
            optional_keys -= base_required

            base_optional = base.__dict__.get('__optional_keys__', set())
            required_keys -= base_optional
            optional_keys |= base_optional
```

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -7769,6 +7769,46 @@ class Cat(Animal):
...
+        class Base1(TypedDict):
+            a: NotRequired[int]
+
+        class Base2(TypedDict):
+            a: Required[str]
+
+        class Child(Base1, Base2):
+            pass
+
+        # Last base wins
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py:7807]: Is it correct that the last base should win? Doesn't that go against how multiple inheritance in Python usually works, where earlier bases in the `__bases__` tuple generally have priority?

```pycon
Python 3.13.0a2+ (heads/main:e9d1360c9a, Nov 24 2023, 11:23:45) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo:
...     x = 1
...
>>> class Bar:
...     x = 2
...
>>> class Baz(Foo, Bar): pass
...
>>> Baz.x
1
```

**Review by AlexWaygood — COMMENTED** [review]
Thanks, I agree that this is a problem that's worth fixing.

> I don't think this change is required; see [#112509 (comment)](https://github.com/python/cpython/issues/112509#issuecomment-1831423442)

I actually think whether or not static type checkers should consider this behaviour illegal according to PEP-589 has little bearing on what we should do at runtime here. I don't want us to start raising a `TypeError` if a `TypedDict` incompatibly overrides a specific key to make it `Required` when the subclass specified the same key as being `NotRequired`. Doing that now would break backwards compatibility at runtime, and I don't think there's a strong motivation for doing so: it's the job of type checkers to catch this kind of error and emit warnings about it. So, if we've ruled out raising a `TypeError` in this kind of situation, we just have to do the thing at runtime that makes the most sense -- and the status quo doesn't really make any sense, I don't think :)

File: Lib/test/test_typing.py (hunk 2 above)
Comment (JelleZijlstra) [reply Lib/test/test_typing.py:7807]: It's arguably not correct as you say, but it's the current behavior for `__annotations__` and changing that behavior seems difficult.

```
>>> class A(TypedDict):
...     a: int
... 
>>> class B(TypedDict):
...     a: str
... 
>>> class C(A, B): pass
... 
>>> C.__annotations__
{'a': <class 'str'>}
```

Comment (AlexWaygood) [reply Lib/test/test_typing.py:7807]: Oof, that seems unfortunate. As you say, though, better to be internally consistent for now, I guess!

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**